internal/net/call/call_test.go

```diff
@@ -1093,6 +1174,74 @@ func TestRefreshDraining(t *testing.T) {
 	}
 }
 
+// TestUnhealthyChannels attempts to use a resolver that returns a mix of
+// healthy and unhealthy endpoints and verifies that calls succeed by being
+// routed to the healthy backend.
+func TestInitiallyUnhealthyEndpoint(t *testing.T) {
+	t.Skip("TODO(sanjay): Implement unhealthy channel avoidance")
+
+	// To fix this, we will avoid connections on which an initial
+	// health-check has not completed. There are some options here:
+	//
+	// 1. We let the balancer Pick() stay as it is currently where it does
+	// not know about connections, just endpoints. After a Pick(), if the
+	// connection does not exist or the health-check hasn't completed, we
+	// Pick() from the subset of connections with healthy endpoints.
+	//
+	// 2. We eagerly connect to any endpoint handed to us by the Resolver.
+	// Balancer is passed a connection only after the initial hand-shake
+	// has completed. Balancer.Pick() returns a connection, not an
+	// endpoint.
+	//
+	// 3. We return a special error to startCall if we are attempting to
+	// use a connection that hasn't completed its initial health
+	// check. startCall retries after a delay if it sees this error.
+	//
+	// Option 2 seems the best? Its main potential downside is that we may
+	// make too many connections. However we can avoid having too many
+	// connections when there are many clients and many servers by
+	// implementing backend subsetting inside the Resolver.
```

Comment:
Option 2 seems the best to me as well. 

Just summarizing to make sure I understand. A `Resolver` returns a set of endpoints. A client eagerly establishes connections to these endpoints and performs health checks against them. Whenever the set of healthy connections changes, the client updates the `Balancer` with them. Client requests are only made on the healthy connections picked by a `Balancer`. If a request is pending on a connection and we think the connection is unhealthy, we abort the call if it is not retriable or retry it on another connection if it is retriable. Am I understanding correctly?

Reply:
Yes. Though initially, I was imagining we just do a single health-check at connection setup time, which will simplify the implementation.

Reply:
Got it, sounds great to me!